src/network/watchdog.go

```diff
@@ -0,0 +1,137 @@
+package network
+
+import (
+	"fmt"
+	"net"
+	"runtime"
+	"time"
+
+	"github.com/libp2p/go-reuseport"
+)
+
+const LISTEN_TIMEOUT = 50
+const BUF_SIZE = 2
+
+/*
+ * Recursively monitors the other nodes.
+ * The closest (forward in the circle) node that is
+ * alive is updated on the updateCurrentNext channel.
+ */
+func monitorNext(
+	nodeID int,
+	nextNodeID int,
+	numNodes int,
+	basePort int,
+	selfDestruct chan bool,
+	updateCurrentNext chan int,
+) {
+	var prevNodeID int
+	hasSubroutine := false
+
+	if nodeID == 0 {
+		prevNodeID = numNodes - 1
+	} else {
+		prevNodeID = nodeID - 1
+	}
+
+	destroySubroutine := make(chan bool)
+
+	pc, err := reuseport.ListenPacket("udp4", fmt.Sprintf(":%d", basePort+nextNodeID))
```

Comment:
Should the "pc" variable be renamed for better clarity? If you are not familiar with networking shortcuts, the purpose of the variable is not clear.